internal/cmd/commands/daemon/version_interceptor.go

```diff
@@ -0,0 +1,32 @@
+package daemon
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/boundary/internal/util"
+	"github.com/hashicorp/boundary/version"
+)
+
+const (
+	VersionHeaderKey = "boundary_version"
+)
+
+// versionEnforcement is an interceptor which, if the boundary version is included
+// in a request, enforces that it matches the version of the daemon currently
+// running. If no version is provided, the inteceptor passes the request through.
+func versionEnforcement(h http.Handler) http.Handler {
+	if util.IsNil(h) {
+		return nil
+	}
+
+	needVer := version.Get().VersionNumber()
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		gotVer := r.Header.Get(VersionHeaderKey)
+		if gotVer != "" && needVer != gotVer {
+			writeError(w, fmt.Sprintf("version mismatch between requester (%q) and server (%q)", gotVer, needVer), http.StatusBadRequest)
```

Comment:
This will presumably be seen on a users terminal, could we add a call to action to the error? Something like "You may need to restart your daemon" (presumably this happens immediately after a user upgrades their CLI).